File: python/src/onetool/testing/world_builder.py
from __future__ import annotations

from ..memory import SimpleMemory
from ..runtime import create_agent_cli
from ..types import ToolAdapters
from ..utils import parent_path
from ..vfs.interface import VFS
from ..vfs.memory_vfs import MemoryVFS
from .adapters import DemoFetch, DemoSearch
from .scenario import WorldSpec

async def build_world(world: WorldSpec, **runtime_options: object):
    options = dict(runtime_options)
    vfs = _resolve_vfs(options.get("vfs"))
    memory = _resolve_memory(options.get("memory"))

    if world.files is not None:
        for file_path, content in sorted(world.files.items()):
            await vfs.mkdir(parent_path(file_path), True)
            data = content if isinstance(content, bytes) else content.encode("utf-8")
            await vfs.write_bytes(file_path, data, False)

    if world.memory is not None:
        for entry in world.memory:
            memory.store(entry)

    options.setdefault("vfs", vfs)
    options.setdefault("memory", memory)
    options.setdefault("adapters", _build_adapters(world))
    if world.output_limits is not None and "output_limits" not in options:
        options["output_limits"] = world.output_limits

    return await create_agent_cli(**options)


def _build_adapters(world: WorldSpec) -> ToolAdapters:
    adapters = ToolAdapters()

    if world.search_docs is not None:
        adapters.search = DemoSearch(world.search_docs)
    if world.fetch_resources is not None:
        adapters.fetch = DemoFetch(world.fetch_resources)

    return adapters


def _resolve_vfs(value: object) -> VFS:
    if value is None:
        return MemoryVFS()
    return value


def _resolve_memory(value: object) -> SimpleMemory:
    if value is None:
        return SimpleMemory()
    return value

File: src/testing/world-builder.ts
import { SimpleMemory } from '../memory.js';
import { createAgentCLI, type AgentCLI } from '../runtime.js';
import type { ToolAdapters } from '../types.js';
import { textEncoder } from '../types.js';
import { parentPath } from '../utils.js';
import { MemoryVFS } from '../vfs/memory-vfs.js';
import { DemoFetch, DemoSearch } from './adapters.js';
import type { WorldSpec } from './scenario.js';
import { compareCLocaleText } from '../utils.js';

export async function buildWorld(world: WorldSpec): Promise<AgentCLI> {
  const vfs = new MemoryVFS();
  const memory = new SimpleMemory();

  if (world.files !== undefined) {
    const fileEntries = Object.entries(world.files).sort(function ([left], [right]) {
      return compareCLocaleText(left, right);
    });

    for (const [filePath, content] of fileEntries) {
      await vfs.mkdir(parentPath(filePath), true);
      const bytes = content instanceof Uint8Array ? new Uint8Array(content) : textEncoder.encode(content);
      await vfs.writeBytes(filePath, bytes, false);
    }
  }

  if (world.memory !== undefined) {
    for (const entry of world.memory) {
      memory.store(entry);
    }
  }

  const options = {
    vfs,
    memory,
    adapters: buildAdapters(world),
  };

  if (world.outputLimits === undefined) {
    return createAgentCLI(options);
  }

  return createAgentCLI({
    ...options,
    outputLimits: world.outputLimits,
  });
}

function buildAdapters(world: WorldSpec): ToolAdapters {
  const adapters: ToolAdapters = {};

  if (world.searchDocs !== undefined) {
    adapters.search = new DemoSearch(world.searchDocs);
  }
  if (world.fetchResources !== undefined) {
    adapters.fetch = new DemoFetch(world.fetchResources);
  }

  return adapters;
}
